src/server/service/user.ts

```diff
@@ -30,7 +30,51 @@ export default class UserService implements IUserService {
     return hash;
   }
 
-  async create(dto: CreateUserDTO): Promise<User> {
+  private checkRoles(role: string): void {
+    const ROLES = [
+      Role.Admin as string,
+      Role.Manager as string,
+      Role.User as string,
+    ];
+
+    if (!ROLES.includes(role)) {
+      throw new InvalidUserRole();
+    }
+  }
+
+  private async checkPermission(email: string, role: Role) {
+    const user = await UserModel.findOne({ email });
+
+    if (!user) {
+      throw new Error(`User with email "${email}" does not exists`);
+    }
+
+    if (user.role === Role.Admin) {
+      if (role !== Role.Manager) {
+        throw new Error(
+          'Admin user does not have permission for creating this user type',
+        );
+      }
+    }
+
+    if (user.role === Role.Manager) {
+      if (role !== Role.User) {
+        throw new Error(
+          'Manager user does not have permission for creating this user type',
+        );
+      }
+    }
+
+    if (user.role === Role.User) {
+      throw new Error(
+        'User does not have permission to create others user type',
+      );
+    }
+  }
+
+  async create(currentUser: User, dto: CreateUserDTO): Promise<User> {
+    this.checkRoles(dto.role);
+    await this.checkPermission(currentUser.role, dto.role);
```

Comment:
I would expand this here, and this logic belongs to this operation and no others